File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/PredicateMatchingRewriter.java
package com.codepoetics.raffia.streaming.rewriting;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.operations.Updater;
import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

final class PredicateMatchingRewriter<T extends BasketWriter<T>> extends StreamingRewriter<T> {

  PredicateMatchingRewriter(T target, FilteringWriter<T> parent, Updater itemUpdater) {
    super(target, parent, itemUpdater);
  }

  @Override
  public FilteringWriter<T> beginObject() {
    return WeavingRewriter.weavingObject(target, this, updater);
  }

  @Override
  public FilteringWriter<T> beginArray() {
    return WeavingRewriter.weavingArray(target, this, updater);
  }

  @Override
  public FilteringWriter<T> end() {
    return parent.advance(target.end());
  }

  @Override
  public FilteringWriter<T> key(String key) {
    return advance(target.key(key));
  }

  @Override
  public FilteringWriter<T> add(String value) {
    return updated(Basket.ofString(value));
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    return updated(Basket.ofNumber(value));
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    return updated(Basket.ofBoolean(value));
  }

  @Override
  public FilteringWriter<T> addNull() {
    return updated(Basket.ofNull());
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/PredicateMatchingRewriter.kt
package com.codepoetics.raffia.streaming.rewriting

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.java.api.Updater
import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

internal class PredicateMatchingRewriter<T : BasketWriter<T>>(target: T, parent: FilteringWriter<T>, itemUpdater: Updater) : StreamingRewriter<T>(target, parent, itemUpdater) {

    override fun beginObject(): FilteringWriter<T> {
        return WeavingRewriter.weavingObject(target, this, updater!!)
    }

    override fun beginArray(): FilteringWriter<T> {
        return WeavingRewriter.weavingArray(target, this, updater!!)
    }

    override fun end(): FilteringWriter<T> {
        return parent!!.advance(target.end())
    }

    override fun key(key: String): FilteringWriter<T> {
        return advance(target.key(key))
    }

    override fun add(value: String): FilteringWriter<T> {
        return updated(Basket.ofString(value))
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        return updated(Basket.ofNumber(value))
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        return updated(Basket.ofBoolean(value))
    }

    override fun addNull(): FilteringWriter<T> {
        return updated(Basket.ofNull())
    }
}
